File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int,int> umap;
        for(int i=0;i<nums1.size();i++){
            umap[nums1[i]]=i;
        }

        stack<int> stack;
        vector<int> ans(nums1.size(),-1);
        for(int i=0;i<nums2.size();i++){
            while(stack.size() && stack.top()<nums2[i]){
                if(umap.count(stack.top()))
                {
                    int n1idx=umap[stack.top()];
                    ans[n1idx]=nums2[i];
                }
                stack.pop();
            }
            stack.push(nums2[i]);
        }
        return ans;
    }
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.py
from collections import defaultdict

class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        freq=defaultdict(int)
        st = [] 
        for i in range(len(nums2)): 
            while st and nums2[st[-1]]<nums2[i]:
                val,idx = nums2[st[-1]],st[-1] 
                st.pop() 
                # print(val,idx)
                freq[val]=nums2[i] 
            st.append(i)    
        res = []
        for val in nums1: 
            if val in freq:
                res.append(freq[val])
            else: 
                res.append(-1)
        return res